File: src/engine/solve.ts
import type { Answer, FlatPuzzle } from "./types.ts";
import { letterIdx } from "./types.ts";
import { deduce } from "./deduce.ts";
import type { DeduceAction } from "./deduce.ts";
import { lookahead } from "./lookahead.ts";
import { checkAnswerValidity } from "./check-validity.ts";

export function solvePuzzle(fp: FlatPuzzle): (Answer | null)[] {
  const n = fp.n;
  const phantomMask = 0b11111 & ~((1 << fp.optionCount) - 1);
  const answers: (Answer | null)[] = new Array(n).fill(null);
  const eliminated: number[] = new Array(n).fill(phantomMask);

  for (let iter = 0; iter < n * 30; iter++) {
    if (answers.slice(0, n).every((a) => a != null)) break;

    const drs = deduce(fp, answers, eliminated);
    if (drs.length > 0) {
      for (const dr of drs) applyAction(dr.action, answers, eliminated);
      continue;
    }

    const lr = lookahead(fp, answers, eliminated);
    if (lr) {
      eliminated[lr.eliminateQi] |= 1 << lr.eliminateOi;
      continue;
    }

    break;
  }

  return answers;
}

export type SolveOutcome = "solved" | "stuck";

export function checkSolvable(fp: FlatPuzzle): SolveOutcome {
  const n = fp.n;
  const phantomMask = 0b11111 & ~((1 << fp.optionCount) - 1);
  const answers: (Answer | null)[] = new Array(n).fill(null);
  const eliminated: number[] = new Array(n).fill(phantomMask);

  for (let iter = 0; iter < n * 30; iter++) {
    if (answers.slice(0, n).every((a) => a != null)) return "solved";

    const drs = deduce(fp, answers, eliminated);
    if (drs.length > 0) {
      for (const dr of drs) applyAction(dr.action, answers, eliminated);
      continue;
    }

    const lr = lookahead(fp, answers, eliminated);
    if (lr) {
      eliminated[lr.eliminateQi] |= 1 << lr.eliminateOi;
      continue;
    }

    break;
  }

  return answers.slice(0, n).every((a) => a != null) ? "solved" : "stuck";
}

export function checkPuzzleSolved(
  fp: FlatPuzzle,
  answers: (Answer | null)[],
  eliminated: number[],
): boolean {
  const n = fp.n;
  for (let i = 0; i < n; i++) {
    if (answers[i] == null) return false;
    if (checkAnswerValidity(fp, answers, eliminated, i) !== "valid") return false;
  }
  return true;
}

function applyAction(action: DeduceAction, answers: (Answer | null)[], eliminated: number[]): void {
  if (action.type === "force") {
    const oi = letterIdx(action.answer);
    eliminated[action.qi] = 0b11111 ^ (1 << oi);
    answers[action.qi] = action.answer;
  } else if (action.type === "eliminateMulti") {
    for (let i = 0; i < eliminated.length; i++) {
      if ((action.questionMask >> i) & 1) eliminated[i] |= action.optionMask;
    }
  } else if (action.type === "eliminate") {
    eliminated[action.qi] |= 1 << action.oi;
  }
}

File: rust/src/solve.rs
use crate::check_validity::{Validity, check_answer_validity};
use crate::deduce::{DeduceAction, deduce};
use crate::lookahead::lookahead;
use crate::types::*;

#[derive(Debug)]
pub enum SolveOutcome {
    Solved,
    Stuck,
}

pub fn check_solvable(fp: &FlatPuzzle) -> SolveOutcome {
    let n = fp.n;
    let mut answers = [None; MAX_N];
    let mut eliminated = [0u8; MAX_N];

    for _ in 0..n * 30 {
        if (0..n).all(|i| answers[i].is_some()) {
            return SolveOutcome::Solved;
        }

        if let Some(dr) = deduce(fp, &answers, &eliminated) {
            apply_action(&dr.action, &mut answers, &mut eliminated);
            continue;
        }

        if let Some(lr) = lookahead(fp, &answers, &eliminated) {
            eliminated[lr.eliminate_qi] |= 1 << lr.eliminate_oi;
            continue;
        }

        break;
    }

    if (0..n).all(|i| answers[i].is_some()) {
        SolveOutcome::Solved
    } else {
        SolveOutcome::Stuck
    }
}

pub fn check_puzzle_solved(
    fp: &FlatPuzzle,
    answers: &[Option<Answer>; MAX_N],
    eliminated: &[u8; MAX_N],
) -> bool {
    let n = fp.n;
    (0..n).all(|i| {
        answers[i].is_some() && check_answer_validity(fp, answers, eliminated, i) == Validity::Valid
    })
}

fn apply_action(
    action: &DeduceAction,
    answers: &mut [Option<Answer>; MAX_N],
    eliminated: &mut [u8; MAX_N],
) {
    match *action {
        DeduceAction::Force { qi, answer } => {
            eliminated[qi] = 0b11111 ^ (1 << answer.idx());
            answers[qi] = Some(answer);
        }
        DeduceAction::Eliminate { qi, oi } => {
            eliminated[qi] |= 1 << oi;
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Value;

    #[test]
    fn test_shared_solve() {
        let json_str =
            std::fs::read_to_string("../tests/solve.json").expect("can't read tests/solve.json");
        let suite: Value = serde_json::from_str(&json_str).unwrap();
        let tests = suite["tests"].as_array().unwrap();

        let mut passed = 0;
        let mut failed = 0;

        for test in tests {
            if test.get("section").is_some() {
                continue;
            }
            let name = test["name"].as_str().unwrap();
            let expect = test["expect"].as_str().unwrap();

            let fp = crate::parse_puzzle(&test["puzzle"]);
            let fp = match fp {
                Some(fp) => fp,
                None => {
                    eprintln!("SKIP: {name}: parse failed");
                    continue;
                }
            };

            let outcome = check_solvable(&fp);
            let got = match outcome {
                SolveOutcome::Solved => "solved",
                SolveOutcome::Stuck => "stuck",
            };

            if got == expect {
                passed += 1;
            } else {
                failed += 1;
                eprintln!("FAIL: {name}");
                eprintln!("  expected: {expect}");
                eprintln!("  got:      {got}");
            }
        }

        eprintln!("{passed}/{} passed", passed + failed);
        assert_eq!(failed, 0, "{failed} test(s) failed");
    }
}
